sam.c

```diff
@@ -1924,13 +1938,655 @@ int sam_parse1(kstring_t *s, sam_hdr_t *h, bam1_t *b)
+                fd->curr_bam = NULL;
+            sp_bams *n = b->next;
+            int i;
+            for (i = 0; i < b->abams; i++)
+                if (b->bams[i].data)
+                    free(b->bams[i].data);
+            free(b->bams);
+            free(b);
+            b = n;
+        }
+
+        if (fd->curr_bam) {
+            int i;
+            for (i = 0; i < fd->curr_bam->abams; i++)
+                if (fd->curr_bam->bams[i].data)
+                    free(fd->curr_bam->bams[i].data);
+            free(fd->curr_bam->bams);
+            free(fd->curr_bam);
+        }
+
+        // Decrement counter by one, maybe destroying too.
+        // This is to permit the caller using bam_hdr_destroy
+        // before sam_close without triggering decode errors
+        // in the background threads.
+        bam_hdr_destroy(fd->h);
+    }
+
+    free(fp->state);
+    fp->state = NULL;
+    return ret;
+}
+
+// Run from one of the worker threads.
+// Convert a passed in array of lines to array of BAMs, returning
+// the result back to the thread queue.
+void *sam_parse_worker(void *arg) {
+    sp_lines *gl = (sp_lines *)arg;
+    sp_bams *gb = NULL;
+    char *lines = gl->data;
+    int i;
+    bam1_t *b;
+    SAM_state *fd = gl->fd;
+
+    // Use a block of BAM structs we had earlier if available.
+    pthread_mutex_lock(&fd->lines_m);
+    if (fd->bams) {
+        gb = fd->bams;
+        fd->bams = gb->next;
+    }
+    pthread_mutex_unlock(&fd->lines_m);
+
+    if (gb == NULL) {
+        gb = calloc(1, sizeof(*gb));
+        if (!gb) {
+            return NULL;
+        }
+        gb->abams = 100;
+        gb->bams = b = calloc(gb->abams, sizeof(*b));
+        if (!gb->bams) {
+            free(gb);
+            return NULL;
+        }
+        gb->nbams = 0;
+    }
+    gb->serial = gl->serial;
+    gb->next = NULL;
+
+    b = (bam1_t *)gb->bams;
+    if (!b)
+        return NULL;
+
+    i = 0;
+    char *cp = lines, *cp_end = lines + gl->data_size;
+    while (cp < cp_end) {
+        if (i >= gb->abams) {
+            int old_abams = gb->abams;
+            gb->abams *= 2;
+            b = (bam1_t *)realloc(gb->bams, gb->abams*sizeof(bam1_t));
+            if (!b)
+                return NULL;
+            memset(&b[old_abams], 0, (gb->abams - old_abams)*sizeof(*b));
+            gb->bams = b;
+        }
+
+        // Ideally we'd get sam_parse1 to return the number of
+        // bytes decoded and to be able to stop on newline as
+        // well as \0.
+        //
+        // We can then avoid the additional strchr loop.
+        // It's around 6% of our CPU cost, albeit threadable.
+        //
+        // However this is an API change so for now we copy.
+
+        char *nl = strchr(cp, '\n');
+        nl = nl ? nl : cp_end;
+        if (*nl) *nl++ = '\0';
+        kstring_t ks = {nl-cp, gl->alloc, cp};
+        if (sam_parse1(&ks, fd->h, &b[i]) < 0) {
+            sam_state_err(fd, EIO);
+            return NULL;
```

Comment:
This could leak a lot of memory on error:

- `sp_lines *gl` passed in as input
- `gl->data`
- `sp_bams *gb`
- `gb->bams`
- `gb->bams[i]->data`

All these `return NULL`s should go to some clean-up code instead, which should either free the data, or possibly put `gl` and `gb` on to the appropriate free lists in the `SAM_state` so they can be reused or (more likely) cleaned up later.